### PR TITLE
Loop over read for readIntoBuffer

### PR DESCRIPTION
**Pull request by xf6wang**
Loop over the read to ensure that we have really read everything we can. 

@bryancall can you take a look?

### CODE REVIEW

File: lib/ts/MatcherUtils.cc [hunk 1]
```diff
@@ -70,30 +72,50 @@ readIntoBuffer(const char *file_path, const char *module_name, int *read_size_pt
...
+  buf = file_buf; // working pointer
+
+  // loop over read, trying to read in as much as we can each time.
+  while (file_size > read_size) {
+    ret = read(fd, buf, file_size - read_size);
+    if (ret <= 0) {
+      break;
+    }
+
+    buf += ret;
```
Comment (xf6wang) [inline lib/ts/MatcherUtils.cc]: is this ok or should I be doing something like `buf = (uint8_t *)buf + ret;`. `buf` is a `char*` but `ret` is number of bytes read.

Comment (shinrich) [reply lib/ts/MatcherUtils.cc]: That should be fine pointer arithmetic.

**Comment by shinrich** [comment]
[approve ci]

File: lib/ts/MatcherUtils.cc [hunk 2]
```diff
@@ -70,30 +72,50 @@ readIntoBuffer(const char *file_path, const char *module_name, int *read_size_pt
...
+  file_size = file_info.st_size; // number of bytes in file
+
+  if (file_size < 0) {
     Error("%s Can not get correct file size for %s file : %" PRId64 "", module_name, file_path, (int64_t)file_info.st_size);
     close(fd);
     return nullptr;
   }
+
+  ink_assert(file_size >= 0);
+  ink_assert(file_size == file_info.st_size);
```
Comment (bryancall) [inline lib/ts/MatcherUtils.cc]: This isn't necessary since you are doing an assignment right above.

File: lib/ts/MatcherUtils.cc [hunk 3]
```diff
@@ -70,30 +72,50 @@ readIntoBuffer(const char *file_path, const char *module_name, int *read_size_pt
...
   // Allocate a buffer large enough to hold the entire file
   //   File size should be small and this makes it easy to
   //   do two passes on the file
-  file_buf = (char *)ats_malloc(file_info.st_size + 1);
+  file_buf = (char *)ats_malloc(file_size + 1);
   // Null terminate the buffer so that string operations will work
-  file_buf[file_info.st_size] = '\0';
+  file_buf[file_size] = '\0';
+
+  ret = 0;
```
Comment (bryancall) [inline lib/ts/MatcherUtils.cc]: ret should be assigned at declaration.  I would move the declaration down here.

File: lib/ts/MatcherUtils.cc (hunk 2 above)
Comment (xf6wang) [reply lib/ts/MatcherUtils.cc]: removed.

File: lib/ts/MatcherUtils.cc (hunk 3 above)
Comment (xf6wang) [reply lib/ts/MatcherUtils.cc]: gotcha. updated.

**Comment by zwoop** [comment]
[approve ci]

**Comment by zwoop** [comment]
Thanks for this fix!

**Pull request merged**